tests/bsim/bluetooth/host/l2cap/multilink_peripheral/src/dut.c

```diff
@@ -0,0 +1,337 @@
+{
+	char addr[BT_ADDR_LE_STR_LEN];
+
+	bt_addr_le_to_str(bt_conn_get_dst(conn), addr, sizeof(addr));
+
+	LOG_DBG("Disconnected from %s (reason 0x%02x)", addr, reason);
+}
+
+static struct bt_conn_cb peripheral_cb = {
+	.connected = acl_connected,
+	.disconnected = acl_disconnected,
+};
+
+static void increment(struct bt_conn *conn, void *user_data)
+{
+	int *conn_count = user_data;
+
+	(*conn_count)++;
+}
+
+static bool have_free_conn(void)
+{
+	int conn_count = 0;
+
+	bt_conn_foreach(BT_CONN_TYPE_LE, increment, &conn_count);
+
+	return conn_count < CONFIG_BT_MAX_CONN;
+}
+
+static const struct bt_data ad[] = {
+	BT_DATA_BYTES(BT_DATA_FLAGS, (BT_LE_AD_GENERAL | BT_LE_AD_NO_BREDR)),
+	BT_DATA(BT_DATA_NAME_COMPLETE, DUT_NAME, sizeof(DUT_NAME) - 1),
+};
+
+static void start_advertising(void)
+{
+	int err;
+
+	if (!have_free_conn() || IS_FLAG_SET(ADVERTISING)) {
+		return;
+	}
+
+	LOG_DBG("starting advertiser");
+
+	err = bt_le_adv_start(BT_LE_ADV_CONN_ONE_TIME, ad, ARRAY_SIZE(ad), NULL, 0);
+	TEST_ASSERT(!err, "Advertising failed to start (err %d)", err);
+
+	SET_FLAG(ADVERTISING);
+}
+
+static bool all_data_transferred(void)
+{
+	int total_sdu_count = 0;
+
+	for (int i = 0; i < ARRAY_SIZE(contexts); i++) {
+		total_sdu_count += contexts[i].sdu_count;
+	}
+
+	TEST_ASSERT(total_sdu_count <= (SDU_NUM * CONFIG_BT_MAX_CONN),
+		    "🎶 too many SDUs, too many SDUs 🎶");
+
+	return total_sdu_count == (SDU_NUM * CONFIG_BT_MAX_CONN);
+}
+
+void entrypoint_dut(void)
+{
+	/* Test purpose:
+	 *
+	 * For a peripheral device (DUT) that has multiple ACL connections to
+	 * central devices: Verify that the data streams on one connection are
+	 * not affected by one of the centrals going out of range or not
+	 * responding.
+	 *
+	 * Three devices:
+	 * - `dut`: sends L2CAP packets to p0 and p1
+	 *
+	 * DUT (in a loop):
+	 * - advertise as connectable
+	 * - [acl connected]
+	 * - establish L2CAP channel
+	 * - [l2 connected]
+	 * - send L2CAP data until ACL disconnected or SDU_NUM SDUs reached
+	 *
+	 * p0/1/2 (in a loop):
+	 * - scan & connect ACL
+	 * - [acl connected]
+	 * - [l2cap dynamic channel connected]
+	 * - receive data from DUT
+	 * - disconnect
+	 *
+	 * Verdict:
+	 * - DUT is able to transfer SDU_NUM SDUs to all peers. Data can be
+	 * dropped but resources should not leak, and the transfer should not
+	 * stall.
+	 */
+	int err;
+
+	/* Mark test as in progress. */
+	TEST_START("dut");
+
+	/* Initialize Bluetooth */
+	err = bt_conn_cb_register(&peripheral_cb);
+	TEST_ASSERT(err == 0, "Can't register callbacks (err %d)", err);
+
+	err = bt_enable(NULL);
+	TEST_ASSERT(err == 0, "Can't enable Bluetooth (err %d)", err);
+
+	LOG_DBG("Bluetooth initialized");
+
+	while (!all_data_transferred()) {
+		start_advertising();
+
+		/* L2 channel is opened from conn->connected() */
+		/* L2 data transfer is initiated from l2->connected() */
+		/* L2 data transfer is initiated for next SDU from l2->sent() */
+
+		k_sleep(K_MSEC(10));
```

Comment:
What is the purpose of this sleep?

Reply:
not hammer the CPU checking if we need to start advertising.

Reply:
Ah, right. 

The naming of `start_advertising` threw me off here, because it's actually `start_advertising_if_not_already_advertising` :) 

Had the check
```
	if (!have_free_conn() || IS_FLAG_SET(ADVERTISING)) {
		return;
	}
```

been here in the loop rather than in the function, it would have been more obvious. 

Similar to the other sleep, probably worthwhile to have a comment in the code for why the sleep is here